hydrolib/core/io/net/models.py

```diff
@@ -367,25 +369,167 @@ def __init__(
+        Raises:
+            ValueError: Raised when any of the structure offsets is smaller than zero.
+            ValueError: Raised when any of the structure offsets is greater than the branch length.
+
+        Returns:
+            np.ndarray: The generated branch offsets.
+        """
+        # Generate initial offsets
+        anchor_pts = [0.0, self.length]
+        offsets = self._generate_1d_spacing(anchor_pts, mesh1d_edge_length)
+
+        if structure_offsets is None:
+            return offsets
+
+        # Check the limits
+        if min(structure_offsets) < 0.0:
+            raise ValueError(
+                f"Distance {min(structure_offsets)} is outside the branch range (0.0 - {self.length})."
+            )
+        if max(structure_offsets) > self.length:
+            raise ValueError(
+                f"Distance {max(structure_offsets)} is outside the branch length (0.0 - {self.length})."
+            )
+
+        # Merge limits with start and end of branch
+        limits = [-1e-3] + list(sorted(structure_offsets)) + [self.length + 1e-3]
+
+        # if requested, check if the calculation point are close enough to the structures
+        if max_dist_to_struc is not None:
+            limits: List[float] = self._update_limits(max_dist_to_struc, limits)
+
+        offsets = self._add_nodes_to_segments(
+            offsets, anchor_pts, limits, mesh1d_edge_length
+        )
+
+        return offsets
+
+    def _update_limits(
+        self, max_dist_to_struc: float, limits: List[float]
+    ) -> List[float]:
+        """Update the limits taking into account the maximum distance to a structure.
+
+        Args:
+            max_dist_to_struc (float): The maximum distance from a node to a structure.
+            limits (List[float]): The limits.
+
+        Returns:
+            List[float]: A list with the updated limits.
+        """
+
+        additional = []
+
+        # Skip the first and the last, these are no structures
+        for i in range(1, len(limits) - 1):
+            # if the distance between two limits is large than twice the max distance to structure,
+            # the mesh point will be too far away. Add a limit on the minimum of half the length and
+            # two times the max distance
+            dist_to_prev_limit = limits[i] - (
+                max(additional[-1], limits[i - 1]) if any(additional) else limits[i - 1]
+            )
+            if dist_to_prev_limit > 2 * max_dist_to_struc:
+                additional.append(
+                    limits[i] - min(2 * max_dist_to_struc, dist_to_prev_limit / 2)
+                )
+
+            dist_to_next_limit = limits[i + 1] - limits[i]
+            if dist_to_next_limit > 2 * max_dist_to_struc:
+                additional.append(
+                    limits[i] + min(2 * max_dist_to_struc, dist_to_next_limit / 2)
+                )
+
+        # Join the limits
+        return sorted(limits + additional)
```

Comment:
I feel like you could probably achieve something more concise with a generator and a stack which yields the next item on the stack and then calculates if an additional node needs to be added something like:

```python
# Note that this untested pseudo-ish code

if len(limits) == 0:
    return []
else:
    return list(_generate_limits)

def _generate_limits(...):    
    stack = list(limits)  # Create a copy to not adjust original limits
    while True:
        curr_item = stack[0]
        yield curr_item
        
        stack = stack[1:]
        if len(stack) > 0:
            if dist_to_next_limit := stack[0] - curr_item > 2 * max_dist_to_struc:
                # Add item to the front of the stack so that it gets yielded next time
                # and we perform the calculation from this point next.
                stack[:0] = curr_item + min(2 * max_dist_to_struc, dist_to_next_limit / 2)
        else:
            # No more items left, time to call it quits
            break
```

Do note that the above code can generate more points than one between two limit nodes. I am not sure if that is by design or a potential bug in the above code. If it is by design, then you might want to adjust the code to yield differently (probably directly yield the extra item instead of pushing it back on the stack).

Further note that because you keep incrementing the nodes, and the original limits are already sorted, your output is also sorted.

Reply:
Thanks for the suggestions.  I leave it as is, as the code is not mine and uncovered by tests, so I don't really dare to change it.